src/amocrm/services/orders/order_pusher.py

```
@@ -0,0 +1,107 @@
+        orders_with_lead = [order for order in orders_with_same_user_and_course if hasattr(order, "amocrm_lead")]
+
+        if len(orders_with_lead) == 1:
+            existing_lead = orders_with_lead[0].amocrm_lead
+            self.link_lead_to_new_order(existing_lead=existing_lead)
+            return existing_lead
+
+        if len(orders_with_lead) > 1:
+            raise AmoCRMOrderPusherException("There are duplicates leads for such order with same course and user")
+
+    def link_lead_to_new_order(self, existing_lead: AmoCRMOrderLead) -> None:
+        existing_lead.order = self.order
+        existing_lead.save()
+
+    def update_lead(self) -> None:
+        from amocrm.tasks import update_amocrm_lead
+
+        update_amocrm_lead.delay(order_id=self.order.id)
+
+    def create_lead(self) -> None:
+        from amocrm.tasks import create_amocrm_lead
+
+        create_amocrm_lead.delay(order_id=self.order.id)
+
+    def update_order_in_amocrm(self) -> None:
+        from amocrm.tasks import push_existing_order_to_amocrm
+
+        push_existing_order_to_amocrm.delay(order_id=self.order.id)
+
+    def create_order_in_amocrm(self) -> None:
+        from amocrm.tasks import push_new_order_to_amocrm
+
+        push_new_order_to_amocrm.delay(order_id=self.order.id)
+
+    @staticmethod
+    def order_must_be_pushed(order: Order) -> bool:
+        if order.is_b2b:
+            return False
+        if order.price == 0:
+            return False
+
+        paid_order = Order.objects.filter(user=order.user, course=order.course, paid__isnull=False, unpaid__isnull=True).last()
+        if paid_order is not None and paid_order != order:
+            return False
+
+        return True
+
+    def get_validators(self) -> list[Callable]:
+        return [
+            self.validate_order_with_course,
+        ]
+
+    def validate_order_with_course(self) -> None:
+        if self.order.course is None:
```

А когда это возможно в реальной жизни?

---

Пожалуй, тут это оверхед. Просто хотел на случай, когда будет новый вид продукта, добавить сразу заметную ошибку

---

Думаю, в этом нет смысла. Let it crash.